### PR TITLE
Value to string conversions

### PR DESCRIPTION
**Pull request by ThoSe1990**

### CODE REVIEW

**Comment by kreuzberger** [comment]
Testing on the branch still fails for some examples:
There are ALWAYS problems with - (minus) and comma (german), but this could also be dependent on the current locale?
This is independent from e.g. to be an int,float or in the example a timestamp
i am using locale de_DE on my linux box.

```
      | -123,12   | -999,9999 | "-123,123"    | "-999,9999"        | 
      | "abc"     | -999,9999 | "abc"         | "-999,9999"        | 
      | abc       | -00:00,00 | abc           | "-00:00,00"        | 
```
The last two "words" (abc) give identical results independent from the quotes. i think this is not what the "standard" would expect.

**Comment by kreuzberger** [comment]
These are the results:
```
Scenario Outline: a scenario outline  examples/features/2_scenario_outline.feature:15
[   PASSED    ] When A <word> and <anonymous>  examples/features/2_scenario_outline.feature:16
Value -123,123 is not equal to 123,12 in following step:
Value -999,9999 is not equal to 999,9999 in following step:
[   FAILED    ] Then They will match <expected_word> and <expected_anonymous>  examples/features/2_scenario_outline.feature:17
  With Examples:
  | word     | anonymous | expected_word | expected_anonymous |
  | 123,12   | 999,9999  | -123,123      | -999,9999          |

Scenario Outline: a scenario outline  examples/features/2_scenario_outline.feature:15
[   PASSED    ] When A <word> and <anonymous>  examples/features/2_scenario_outline.feature:16
Value -999,9999 is not equal to 999,9999 in following step:
[   FAILED    ] Then They will match <expected_word> and <expected_anonymous>  examples/features/2_scenario_outline.feature:17
  With Examples:
  | word     | anonymous | expected_word | expected_anonymous |
  | abc      | 999,9999  | abc           | -999,9999          |

Scenario Outline: a scenario outline  examples/features/2_scenario_outline.feature:15
[   PASSED    ] When A <word> and <anonymous>  examples/features/2_scenario_outline.feature:16
Value -00:00,00 is not equal to 00:00,00 in following step:
[   FAILED    ] Then They will match <expected_word> and <expected_anonymous>  examples/features/2_scenario_outline.feature:17
  With Examples:
  | word     | anonymous | expected_word | expected_anonymous |
  | abc      | 00:00,00  | abc           | -00:00,00          |
```

**Comment by ThoSe1990** [comment]
Thanks for testing, I just had a great refactoring here.  I will further test it before merging. Good findings.  I re-opened the issue https://github.com/ThoSe1990/cwt-cucumber/issues/38 and added an explanation here. I think this is going into the right direction. 

> TL;DR: I refactored a lot and came to a different approach. All tests pass and internally I hold all values as std::string. it comes down to the step definition to which type you want to convert. 
--- 
>In earlier implementations I needed the type information, since i parsed features/steps myself. In order to match to the correct type, I used type erasure in cuke::value. So every value was represented as cuke::value and with some type traits I was able to detect if {int} matches an int, and so on and so forth.

> Since this all is now out and we introduced regex to find the steps, I could remove the interface inside cuke::value and every value is now represented as string internally. Only in the finale state when the developer in their step definition access the value, i convert the strings to int, double, etc.

> With this the printing issue resolves, because there is no conversion from string -> float/double and back to string again. (even though it looked good with the format library)

> This is a massive simplification internally and for now all tests pass and it looks very good. I'll play around and sleep a night on it but this looks as a great improvement to me.

> So at this point, thank you so much, for having a detailed look on this project, it improved already a lot.

If there aren't any issues, I'll merge this branch.

**Comment by ThoSe1990** [comment]
just tested with: 
```
  Scenario Outline: a scenario outline  
    When A <word> and <anonymous>
    Then They will match <expected_word> and <expected_anonymous>
    
    Examples:
      | word      | anonymous | expected_word | expected_anonymous |
      | 123.123   | -999.9999 | "123.123"     | "-999.9999"        | 
      | -123,123  | -999,9999 | "-123,123"    | "-999,9999"        | 
      | "abc"     | -999,9999 | "abc"         | "-999,9999"        | 
      | abc       | -00:00,00 | abc           | "-00:00,00"        | 
```

and it passed. 

I'll do some research with the strings inside tables, in this implementation i basically allow unquotet strings in the table, I'll check how other cucumber implementations do this.

**Comment by kreuzberger** [comment]
i have tested with pytest-bdd-ng and there {word} and {} behave identical. Both give the "raw" text back as string. If it contains quotes, the quotes are part of the string.
{string} REQUIRES the quotes, else give parse errors and the result does not contain quotes.
In the cucumber-expression repository under tests there are some examples, unfortunately not for {word}

**Comment by ThoSe1990** [comment]
I see. Do you need it in the exact same behavior? 

So right now I'm removing the quotes from the string and inside a table unquoted strings are allowed, does not matter if its `{string}`, `{word}` or `{}`.  If this is a necessary feature to identify "raw" strings then I'd add some type checks when creating the values.

**Comment by kreuzberger** [comment]
> I see. Thanks a lot for checking out. Do you need it in the exact same behavior?
> 
> So right now I'm removing the quotes from the string and inside a table unquoted strings are allowed, does not matter if its `{string}`, `{word}` or `{}`. If this is a necessary feature to identify "raw" strings then I'd add some type checks when creating the values.

I do not need it in the exact behaviour. Unquoted strings and and handling of {word} {string} should be different for datatables to differ the types, don't you think?

i expect a {word} to fail if the value in the data table could not be mapped, e.g. a raw `ice cream` entry.
i expect a {word} to fail if the value in the data table is quoted `"ice cream"`
I would expect a {string} to work if the data table entry would be `"ice cream"`. 
I think its acceptable if a {string} works with a raw unquoted `ice cream`
I think its acceptable if a {word} returns `bananas` for quoted `"bananas"` entry, but wouldn't this makes the parser more complicated? Returning `"bananas"` would also be ok in this case.

**Comment by ThoSe1990** [comment]
I don't really have a preference here, but anyway after thinking it through, checking the type would make sense, also in order to get proper error messages when mistakes happen (like I define {int} but in the feature file I accidentally write a string...) 

And with this i also can check quotes for strings. Non quotes for words/anonymous and storing their values with and without strings. 

Thanks a lot it helped, thinking it through one more time :)

**Comment by kreuzberger** [comment]
Maybe we think to complicated. And do not consider when to check / validate.

Given a Test Step Implementation `This is a {string}`.

Should work for Scenario outlines defined as
```
Given This is a <example>
Examples:
|example|
|"hello"|
```
And also with

```
Given This is a "<example>"
Examples:
|example|
|hello  |
```

both result in a valid, quoted string

**Comment by ThoSe1990** [comment]
You're right, considering the options, i think a valid approach is to do this in two steps: 
1. search & replace all variables between angle brackets
2. check if the step matches 

and i think this is pretty straightforward, then cells are treated as they were in the statement and this also tackles the issue https://github.com/ThoSe1990/cwt-cucumber/issues/39 because this is the function i actually need here.

**Comment by ThoSe1990** [comment]
so now this all passes and I'd say as expected: 

```
  Scenario Outline: a scenario outline  
    When A <word> and <anonymous>
    Then They will match <expected word> and <expected anonymous>
    
    Examples:
      | word      | anonymous | expected word | expected anonymous |
      | 123.123   | -999.9999 | "123.123"     | "-999.9999"        | 
      | -123,123  | -999,9999 | "-123,123"    | "-999,9999"        | 
      | "abc"     | -999,9999 | ""abc""       | "-999,9999"        | 
      | abc       | -00:00,00 | "abc"         | "-00:00,00"        | 


  Scenario Outline: lets put the quotes in the step ...  
    When A <word> and <anonymous>
    Then They will match "<expected word>" and "<expected anonymous>"
    
    Examples:
      | word      | anonymous | expected word | expected anonymous |
      | 123.123   | -999.9999 | 123.123       | -999.9999          | 
      | "abc"     | -00:00,00 | "abc"         | -00:00,00          | 

```

Also I implemented better step prints in scenario outline (https://github.com/ThoSe1990/cwt-cucumber/issues/39) which now looks like this: 

```
Scenario Outline: a scenario outline  ./examples/features/2_scenario_outline.feature:15
[   PASSED    ] When A -123,123 and -999,9999  ./examples/features/2_scenario_outline.feature:16
[   PASSED    ] Then They will match "-123,123" and "-999,9999"  ./examples/features/2_scenario_outline.feature:17

Scenario Outline: a scenario outline  ./examples/features/2_scenario_outline.feature:15
[   PASSED    ] When A "abc" and -999,9999  ./examples/features/2_scenario_outline.feature:16
[   PASSED    ] Then They will match ""abc"" and "-999,9999"  ./examples/features/2_scenario_outline.feature:17
```

strings: need quotes 
words: a single word 
anonymous: anything 

Mismatches will lead to undefined steps


I'll go over it and check for left overs tomorrow. IMO its looking good. Thanks a lot for paying attention to all the details.
